### ISSUE TITLE
Figures aren't accessible in HTML version

### ISSUE THREAD

**Issue by ethindp** [riscv/riscv-isa-manual#1129]
The HTML version has figures for things like instruction encodings, register layouts, etc., but they appear as just graphics and not in an accessible form (e.g. tables). I don't know if this is a wavedrom issue or not. I can look at the layouts by looking at the wavedrom source, but it's not the most desirable/preferable method.

**Comment by wmat** [comment]
I'm not sure what you mean by accessible form?  The wavedroms are rendered as svg images.

**Comment by ethindp** [comment]
I mean they aren't rendered as, e.g., a table.

**Comment by wmat** [comment]
Just so I'm clear, are you asking for the wavedrom content to be rendered as HTML tables?  If that's what you're after, what benefit would that provide over a svg?  Could you provide a specific example from the spec and what you mean by consuming it?

**Comment by ethindp** [comment]
@wmat The benefit would be that people with screen readers, like myself, would be able to understand the tables. As it currently stands, the figures just say "graphic diagram" when I arrow over them, which provides no helpful information. Looking at the code, the images are being generated as follows:

```html
<div class="imageblock">
<div class="content">
<img src="data:image/svg+xml;base64,PHN2ZyBjbGFzcz0nV2F2ZURyb20nIGhlaWdodD0nNTYnIHByZXNlcnZlQXNwZWN0UmF0aW89J3hNaWRZTWlkIG1lZXQnIHZpZXdCb3g9JzAgMCA4MDAgNTYnIHdpZHRoPSc4MDAnIHhtbG5zPSdodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2Zyc+PGcgZm9udC1mYW1pbHk9J3NhbnMtc2VyaWYnIGZvbnQtc2l6ZT0nMTQnIGZvbnQtd2VpZ2h0PSdub3JtYWwnIHRleHQtYW5jaG9yPSdtaWRkbGUnIHRyYW5zZm9ybT0ndHJhbnNsYXRlKDAuNSwwLjUpJz48ZyB0cmFuc2Zvcm09J3RyYW5zbGF0ZSg0LDIxKSc+PGcgc3Ryb2tlPSdibGFjaycgc3Ryb2tlLWxpbmVjYXA9J3JvdW5kJyBzdHJva2Utd2lkdGg9JzEnPjxsaW5lIHgyPSc3MTUnLz48bGluZSB5Mj0nMzEnLz48bGluZSB4Mj0nNzE1JyB5MT0nMzEnIHkyPSczMScvPjxsaW5lIHgxPSc3MTUnIHgyPSc3MTUnIHkyPSczMScvPjxsaW5lIHgxPSc2OTMnIHgyPSc2OTMnIHkyPSczJy8+PGxpbmUgeDE9JzY5MycgeDI9JzY5MycgeTE9JzMxJyB5Mj0nMjgnLz48bGluZSB4MT0nNjcwJyB4Mj0nNjcwJyB5Mj0nMycvPjxsaW5lIHgxPSc2NzAnIHgyPSc2NzAnIHkxPSczMScgeTI9JzI4Jy8+PGxpbmUgeDE9JzY0OCcgeDI9JzY0OCcgeTI9JzMnLz48bGluZSB4MT0nNjQ4JyB4Mj0nNjQ4JyB5MT0nMzEnIHkyPScyOCcvPjxsaW5lIHgxPSc2MjYnIHgyPSc2MjYnIHkyPSczJy8+PGxpbmUgeDE9JzYyNicgeDI9JzYyNicgeTE9JzMxJyB5Mj0nMjgnLz48bGluZSB4MT0nNjAzJyB4Mj0nNjAzJyB5Mj0nMycvPjxsaW5lIHgxPSc2MDMnIHgyPSc2MDMnIHkxPSczMScgeTI9JzI4Jy8+PGxpbmUgeDE9JzU4MScgeDI9JzU4MScgeTI9JzMnLz48bGluZSB4MT0nNTgxJyB4Mj0nNTgxJyB5MT0nMzEnIHkyPScyOCcvPjxsaW5lIHgxPSc1NTknIHgyPSc1NTknIHkyPSczMScvPjxsaW5lIHgxPSc1MzYnIHgyPSc1MzYnIHkyPSczJy8+PGxpbmUgeDE9JzUzNicgeDI9JzUzNicgeTE9JzMxJyB5Mj0nMjgnLz48bGluZSB4MT0nNTE0JyB4Mj0nNTE0JyB5Mj0nMycvPjxsaW5lIHgxPSc1MTQnIHgyPSc1MTQnIHkxPSczMScgeTI9JzI4Jy8+PGxpbmUgeDE9JzQ5MicgeDI9JzQ5MicgeTI9JzMnLz48bGluZSB4MT0nNDkyJyB4Mj0nNDkyJyB5MT0nMzEnIHkyPScyOCcvPjxsaW5lIHgxPSc0NjknIHgyPSc0NjknIHkyPSczJy8+PGxpbmUgeDE9JzQ2OScgeDI9JzQ2OScgeTE9JzMxJyB5Mj0nMjgnLz48bGluZSB4MT0nNDQ3JyB4Mj0nNDQ3JyB5Mj0nMzEnLz48bGluZSB4MT0nNDI1JyB4Mj0nNDI1JyB5Mj0nMycvPjxsaW5lIHgxPSc0MjUnIHgyPSc0MjUnIHkxPSczMScgeTI9JzI4Jy8+PGxpbmUgeDE9JzQwMicgeDI9JzQwMicgeTI9JzMnLz48bGluZSB4MT0nNDAyJyB4Mj0nNDAyJyB5MT0nMzEnIHkyPScyOCcvPjxsaW5lIHgxPSczODAnIHgyPSczODAnIHkyPSczMScvPjxsaW5lIHgxPSczNTgnIHgyPSczNTgnIHkyPSczJy8+PGxpbmUgeDE9JzM1OCcgeDI9JzM1OCcgeTE9JzMxJyB5Mj0nMjgnLz48bGluZSB4MT0nMzM1JyB4Mj0nMzM1JyB5Mj0nMycvPjxsaW5lIHgxPSczMzUnIHgyPSczMzUnIHkxPSczMScgeTI9JzI4Jy8+PGxpbmUgeDE9JzMxMycgeDI9JzMxMycgeTI9JzMnLz48bGluZSB4MT0nMzEzJyB4Mj0nMzEzJyB5MT0nMzEnIHkyPScyOCcvPjxsaW5lIHgxPScyOTAnIHgyPScyOTAnIHkyPSczJy8+PGxpbmUgeDE9JzI5MCcgeDI9JzI5MCcgeTE9JzMxJyB5Mj0nMjgnLz48bGluZSB4MT0nMjY4JyB4Mj0nMjY4JyB5Mj0nMzEnLz48bGluZSB4MT0nMjQ2JyB4Mj0nMjQ2JyB5Mj0nMycvPjxsaW5lIHgxPScyNDYnIHgyPScyNDYnIHkxPSczMScgeTI9JzI4Jy8+PGxpbmUgeDE9JzIyMycgeDI9JzIyMycgeTI9JzMnLz48bGluZSB4MT0nMjIzJyB4Mj0nMjIzJyB5MT0nMzEnIHkyPScyOCcvPjxsaW5lIHgxPScyMDEnIHgyPScyMDEnIHkyPSczJy8+PGxpbmUgeDE9JzIwMScgeDI9JzIwMScgeTE9JzMxJyB5Mj0nMjgnLz48bGluZSB4MT0nMTc5JyB4Mj0nMTc5JyB5Mj0nMycvPjxsaW5lIHgxPScxNzknIHgyPScxNzknIHkxPSczMScgeTI9JzI4Jy8+PGxpbmUgeDE9JzE1NicgeDI9JzE1NicgeTI9JzMxJy8+PGxpbmUgeDE9JzEzNCcgeDI9JzEzNCcgeTI9JzMnLz48bGluZSB4MT0nMTM0JyB4Mj0nMTM0JyB5MT0nMzEnIHkyPScyOCcvPjxsaW5lIHgxPScxMTInIHgyPScxMTInIHkyPSczJy8+PGxpbmUgeDE9JzExMicgeDI9JzExMicgeTE9JzMxJyB5Mj0nMjgnLz48bGluZSB4MT0nODknIHgyPSc4OScgeTI9JzMnLz48bGluZSB4MT0nODknIHgyPSc4OScgeTE9JzMxJyB5Mj0nMjgnLz48bGluZSB4MT0nNjcnIHgyPSc2NycgeTI9JzMnLz48bGluZSB4MT0nNjcnIHgyPSc2NycgeTE9JzMxJyB5Mj0nMjgnLz48bGluZSB4MT0nNDUnIHgyPSc0NScgeTI9JzMnLz48bGluZSB4MT0nNDUnIHgyPSc0NScgeTE9JzMxJyB5Mj0nMjgnLz48bGluZSB4MT0nMjInIHgyPScyMicgeTI9JzMnLz48bGluZSB4MT0nMjInIHgyPScyMicgeTE9JzMxJyB5Mj0nMjgnLz48L2c+PGc+PGc+PHJlY3QgZmllbGQ9J29wY29kZScgaGVpZ2h0PSczMScgc3R5bGU9J2ZpbGwtb3BhY2l0eTowLjEnIHdpZHRoPScxNTYnIHg9JzU1OScvPjxyZWN0IGZpZWxkPSdyZCcgaGVpZ2h0PSczMScgc3R5bGU9J2ZpbGwtb3BhY2l0eTowLjE7ZmlsbDpoc2woMCwxMDAlLDUwJSknIHdpZHRoPScxMTInIHg9JzQ0NycvPjxyZWN0IGZpZWxkPSdmdW5jdDMnIGhlaWdodD0nMzEnIHN0eWxlPSdmaWxsLW9wYWNpdHk6MC4xJyB3aWR0aD0nNjcnIHg9JzM4MCcvPjxyZWN0IGZpZWxkPSdyczEnIGhlaWdodD0nMzEnIHN0eWxlPSdmaWxsLW9wYWNpdHk6MC4xO2ZpbGw6aHNsKDE3MCwxMDAlLDUwJSknIHdpZHRoPScxMTInIHg9JzI2OCcvPjxyZWN0IGZpZWxkPSdyczInIGhlaWdodD0nMzEnIHN0eWxlPSdmaWxsLW9wYWNpdHk6MC4xO2ZpbGw6aHNsKDE3MCwxMDAlLDUwJSknIHdpZHRoPScxMTInIHg9JzE1NicvPjxyZWN0IGZpZWxkPSdmdW5jdDcnIGhlaWdodD0nMzEnIHN0eWxlPSdmaWxsLW9wYWNpdHk6MC4xJyB3aWR0aD0nMTU2Jy8+PC9nPjxnIHRyYW5zZm9ybT0ndHJhbnNsYXRlKDExLC0xMSknPjxnIHRyYW5zZm9ybT0ndHJhbnNsYXRlKDY5MyknPjx0ZXh0IHk9JzYnPjA8L3RleHQ+PC9nPjxnIHRyYW5zZm9ybT0ndHJhbnNsYXRlKDU1OSknPjx0ZXh0IHk9JzYnPjY8L3RleHQ+PC9nPjxnIHRyYW5zZm9ybT0ndHJhbnNsYXRlKDUzNiknPjx0ZXh0IHk9JzYnPjc8L3RleHQ+PC9nPjxnIHRyYW5zZm9ybT0ndHJhbnNsYXRlKDQ0NyknPjx0ZXh0IHk9JzYnPjExPC90ZXh0PjwvZz48ZyB0cmFuc2Zvcm09J3RyYW5zbGF0ZSg0MjUpJz48dGV4dCB5PSc2Jz4xMjwvdGV4dD48L2c+PGcgdHJhbnNmb3JtPSd0cmFuc2xhdGUoMzgwKSc+PHRleHQgeT0nNic+MTQ8L3RleHQ+PC9nPjxnIHRyYW5zZm9ybT0ndHJhbnNsYXRlKDM1OCknPjx0ZXh0IHk9JzYnPjE1PC90ZXh0PjwvZz48ZyB0cmFuc2Zvcm09J3RyYW5zbGF0ZSgyNjgpJz48dGV4dCB5PSc2Jz4xOTwvdGV4dD48L2c+PGcgdHJhbnNmb3JtPSd0cmFuc2xhdGUoMjQ2KSc+PHRleHQgeT0nNic+MjA8L3RleHQ+PC9nPjxnIHRyYW5zZm9ybT0ndHJhbnNsYXRlKDE1NiknPjx0ZXh0IHk9JzYnPjI0PC90ZXh0PjwvZz48ZyB0cmFuc2Zvcm09J3RyYW5zbGF0ZSgxMzQpJz48dGV4dCB5PSc2Jz4yNTwvdGV4dD48L2c+PGcgdHJhbnNmb3JtPSd0cmFuc2xhdGUoMCknPjx0ZXh0IHk9JzYnPjMxPC90ZXh0PjwvZz48L2c+PGcgdHJhbnNmb3JtPSd0cmFuc2xhdGUoMTEsMTUpJz48ZyB0cmFuc2Zvcm09J3RyYW5zbGF0ZSg2MjYpJz48dGV4dCB5PSc2Jz48dHNwYW4+b3Bjb2RlPC90c3Bhbj48L3RleHQ+PC9nPjxnIHRyYW5zZm9ybT0ndHJhbnNsYXRlKDQ5MiknPjx0ZXh0IHk9JzYnPjx0c3Bhbj5yZDwvdHNwYW4+PC90ZXh0PjwvZz48ZyB0cmFuc2Zvcm09J3RyYW5zbGF0ZSg0MDIpJz48dGV4dCB5PSc2Jz48dHNwYW4+ZnVuY3QzPC90c3Bhbj48L3RleHQ+PC9nPjxnIHRyYW5zZm9ybT0ndHJhbnNsYXRlKDMxMyknPjx0ZXh0IHk9JzYnPjx0c3Bhbj5yczE8L3RzcGFuPjwvdGV4dD48L2c+PGcgdHJhbnNmb3JtPSd0cmFuc2xhdGUoMjAxKSc+PHRleHQgeT0nNic+PHRzcGFuPnJzMjwvdHNwYW4+PC90ZXh0PjwvZz48ZyB0cmFuc2Zvcm09J3RyYW5zbGF0ZSg2NyknPjx0ZXh0IHk9JzYnPjx0c3Bhbj5mdW5jdDc8L3RzcGFuPjwvdGV4dD48L2c+PC9nPjxnIHRyYW5zZm9ybT0ndHJhbnNsYXRlKDExLDM5KScvPjwvZz48ZyB0ZXh0LWFuY2hvcj0nc3RhcnQnPjxnIHRyYW5zZm9ybT0ndHJhbnNsYXRlKDcxOSwxNiknPjx0ZXh0IHk9JzYnPjx0c3Bhbj5SLVR5cGU8L3RzcGFuPjwvdGV4dD48L2c+PC9nPjwvZz48L2c+PC9zdmc+" alt="Diagram" width="800" height="56">
</div>
</div>
```

**Comment by ethindp** [comment]
An alternative would be to include the SVG directly without base64-encoding it and to instead include accessible SVG attributes.

**Comment by ethindp** [comment]
You can find more info about accessible SVGs [here](https://css-tricks.com/accessible-svgs), but in general a table is better and makes interpreting the data much simpler.

**Comment by nick-knight** [comment]
I think it's a reasonable request.

One way forward may be to request WaveDrom to provide some kind of conversion tool that "renders" WaveDrom textual descriptions in a format that is friendly to screen readers. The drawback of this approach is that it would require an enhancement to WaveDrom, a third-party tool. However, this tool is maintained by a kind-hearted RISC-V community member, @drom. Even if he doesn't have the bandwidth to implement this feature, I suspect he would welcome a PR.

Another way forward may be to write "alt text" for each WaveDrom graphic. The drawback of this approach is it adds redundancy to every graphic, meaning extra work (and chances for errors) whenever changes are made to them.

**Comment by ethindp** [comment]
@nick-knight SVGs have accessibility things that WaveDrom could implement as well, including ARIA support. The problem is that to use ARIA properly, one must implement the behaviors that an ARIA role communicates, since ARIA itself does not alter behavior. Of course, as the ARIA authoring guide indicates, no ARIA is better than bad ARIA; in other words, it's better if one simply doesn't need ARIA than one needing to implement it. But the authoring guide for properly using ARIA may be of interest: https://www.w3.org/WAI/ARIA/apg/practices/

**Comment by ethindp** [comment]
Okay, so I have an update on this. To my surprise, AI's image descriptions (at least chatGPT's) are actually quite useful in hacking around this issue. Obviously, this is not an actual fix to the problem, and an actual fix should definitely be worked upon (e.g. maybe implementing the generation of tables instead of diagrams for bitfields/registers/instruction encodings/whatnot). But it's decent at describing the images, even if they lack context (the image description feature is an add-on to my screen reader, and I'm waiting for an update to allow for context-specific descriptions). It's a decent alternative to having to open the wavedrom files alongside the manual. But hopefully we can get a much higher quality accessibility implementation into wavedrom itself.

**Comment by bhanuponguru** [comment]
yes, as a screen reader user, I too face the same issue with accessibility in figures. I too agree with tabular form, I am not knowledged about svg so cannot comment anything about it.

**Comment by Maanvi212006** [comment]
Hi @aswaterman @ethindp ! I’d like to work on this issue if that’s okay.

I’m thinking of starting with an accessibility improvement by adding better alt text and accessible SVG metadata (`<title>`, `<desc>`, ARIA labels) to the generated WaveDrom diagrams so screen readers can provide meaningful descriptions instead of only announcing “graphic diagram”.

If this approach sounds reasonable, could this issue be assigned to me?

**Comment by omasanori** [comment]
Sounds great. For typical instruction encoding and CSR diagrams, we would need a systematic way, as adding markup to hundreds or thousands of diagrams manually is a waste of resources, though.

**Comment by aswaterman** [comment]
That would be a great help.  Let's see a couple examples before we commit to the code style.

**Comment by Maanvi212006** [comment]
Hi! I've worked on a proof-of-concept for the accessibility improvement.

I wrote a script tools/wavedrom_a11y.py in docs-resources that processes .edn files and injects title, desc and ARIA attributes into the generated SVGs so screen readers get meaningful descriptions instead of "graphic diagram".
Branch link : [https://github.com/Maanvi212006/riscv-docs-resources/tree/accessibility/wavedrom-a11y](https://github.com/Maanvi212006/riscv-docs-resources/tree/accessibility/wavedrom-a11y)
wavedrom_a11y.py Path : [riscv-docs-resources/tools/wavedrom_a11y.py](https://github.com/Maanvi212006/riscv-docs-resources/blob/accessibility/wavedrom-a11y/tools/wavedrom_a11y.py)

Example 1 — b-immediate.edn:

<img width="1448" height="272" alt="Image" src="https://github.com/user-attachments/assets/d6365c2e-60db-4bc6-9be6-60ce078fc9bc" />

Example 2 — atomic-mem.edn:

<img width="1437" height="232" alt="Image" src="https://github.com/user-attachments/assets/da439f0a-f989-441e-8a16-a06c80dbdac1" />

Example 3 — c-andi.edn:

<img width="1453" height="253" alt="Image" src="https://github.com/user-attachments/assets/6967ec37-a135-4fd7-b642-aefa64dda4cd" />

The script works systematically across all .edn files without manual edits to each one.
Happy to integrate this into the build pipeline once the approach is approved!

**Comment by omasanori** [comment]
@Maanvi212006 Cool! Few questions:

1. As @ethindp pointed out, the diagrams are currently embedded in the document awfully (from accessibility point of view) like `<img src="data:image/svg+xml;base64,..." alt="Diagram">`. Does it work well with screen readers like NVDA even with this Base64-encoded form?
2. My `src/unpriv/images/bytefield/fprs.edn` is unfortunately one of the most tricky one to your script. Maybe we will need a way to specify the text description manually for such ones. Any ideas?

**Comment by Maanvi212006** [comment]
Hey @omasanori 

**On the base64 issue**:
Yeah, you're absolutely right. When the SVG is embedded as a base64 <img> tag, screen readers can't actually go inside it to read the <title> and <desc> — they only see whatever is in the alt attribute on the <img> tag itself.
So the fix needs to happen at two places — inside the SVG (which the script already does) and also at the AsciiDoctor level where the <img> tag is generated, so the alt attribute gets our generated title too. I'll dig into how AsciiDoctor handles wavedrom block embedding and figure out the right place to hook into that.

**On the bytefield diagrams**:
Totally see the problem — files like fprs.edn that use a LISP-like syntax so the script can't really parse automatically the way it does with WaveDrom JSON.
My idea for handling these is a special comment in the .edn file itself, something like:
`// a11y-desc: Floating-point registers f0-f31, each FLEN bits wide, and fcsr control register, 32 bits wide`
The script would look for this comment and use it as the description when it can't generate one automatically. So the common WaveDrom diagrams get handled automatically, and the trickier ones just need one line added by hand.

Does that sound like a reasonable approach?

**Comment by omasanori** [comment]
Sounds reasonable to me, thanks for answering.

**Comment by wmat** [comment]
I just want to point out that accessibility has long been on the Docs Team's radar and any help with this is greatly appreciated.  Please loop in myself and @kersten1 along the way so we can ensure we apply accessibilty improvements across the RISC-V documentation landscape.  Thanks again.

**Comment by Maanvi212006** [comment]
Thanks @wmat! I'll make sure to tag you and @kersten1 as the work progresses. I'll wait for @aswaterman's and others go-ahead on the approach before opening the PR, and will keep the script generic enough to be reusable across other RISC-V documentation repos too.

**Comment by aswaterman** [comment]
I'll rely on @wmat, @kersten1, and @omasanori to give their feedback on the first few converted figures; I'll chime in later. Thanks again for the help.

**Comment by Maanvi212006** [comment]
The script has been merged into docs-resources. For the next step, I am planning to open a PR here to update the submodule pointer and apply the accessibility fixes to the .edn files in the manual. Does this sound like the right approach? Let me know if there's anything specific you'd like me to address before I do that.
